File: 09.py
#!/usr/bin/env python3
from sys import stdin
from math import prod
from itertools import product

ps   = [(0,-1),(0,1),(-1,0),(1,0)]
rs   = list(map(lambda x: list(map(int, x.strip())), stdin.readlines()))
n,m  = len(rs), len(rs[0])
adj  = lambda i,j: ((a+i,b+j) for a,b in ps if n>a+i>=0 and m>b+j>=0)
v    = lambda i,j: rs[i][j]
comp = lambda i,j: all([v(i,j) < v(a,b) for a,b in adj(i,j)])
lps  = [(i,j) for i,j in product(range(n),range(m)) if comp(i,j)]
B_   = lambda i,j: [(a,b) for a,b in adj(i,j) if v(i,j)<v(a,b) and v(a,b)<9]
Br   = lambda i,j: sum((Br(a,b) for a,b in B_(i,j)),[]) + B_(i,j)
B    = lambda i,j: list(set(Br(i,j)+[(i,j)]))
p1   = sum(1+v(i,j) for i,j in lps)
p2   = sorted(map(lambda x: len(B(*x)), lps))

print (f"Part 1: { p1 }, Part 2: { prod(p2[-3:]) }")

File: 09.scala
#!/usr/bin/env scala
import scala.io.Source.stdin

@main def aoc(): Unit =
  val ps    = List((0,-1),(0,1),(-1,0),(1,0))
  val rs    = stdin.getLines().map { _.map(_.asDigit).toList }.toList
  val (n,m) = (rs.length, rs(0).length)
  val adj   = (i:Int, j:Int) => for { (a,b) <- ps if n>a+i && a+i>=0 && m>b+j && b+j>=0 } yield (a+i,b+j)
  val comp  = (i:Int, j:Int) => adj(i,j).forall(rs(i)(j) < rs(_)(_))
  val lps   = for { i<-(0 to n-1); j<-(0 to m-1) if comp(i,j) } yield (i,j)
  val B_    = (i:Int, j:Int) => adj(i,j).filter { (a,b) => rs(i)(j) < rs(a)(b) && rs(a)(b) < 9 }

  def Br(i:Int, j:Int): List[(Int,Int)] = B_(i,j).flatMap( Br.tupled ) ++ B_(i,j)

  val B     = (i:Int, j:Int) => Br(i,j).appended((i,j)).distinct
  val p2    = lps.map { (i,j)=> B(i,j).length }.sorted

  println(s"Part 1: ${ lps.map(1+rs(_)(_)).sum }, Part 2: ${ p2.takeRight(3).product }")
